### PR TITLE
Global Styles: Add CSS vars for root level settings

### PR DESCRIPTION
**Pull request by scruffian**
## Description
We need a way for blocks to implement the default colors set in root. Often this happens via CSS inheritance but sometimes we want to use the theme colors in a block in a different way, which won't work with inheritance.

This PRs adds CSS variables for root level settings from theme.json:

```
--wp--style--color--background: #0000ff;
 --wp--style--color--text: #ff0000;
 --wp--style--typography--font-size: var(--wp--preset--font-size--normal);
 --wp--style--typography--line-height: var(--wp--custom--line-height--body);
 --wp--style--typography--font-family: var(--wp--preset--font-family--base);
```

This gives blocks a standard way of accessing the default colors for a site which is independent of the theme.

This PR was co-created with @MaggieCabrera 

## How has this been tested?
- Switch to a block theme
- Set colors for the text color and background

## Types of changes
New feature (non-breaking change which adds functionality) 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by nosolosw** [comment]
:wave: I've looked a bit into this and wanted to share some feedback.

For context, we've tried this approach at the beginning of global styles (see https://github.com/WordPress/gutenberg/issues/22296 for a recap) and the issues boiled down to:

- **Opinions**. What if the calendar block had different opinions about defaults? Or how a theme can change the design tokens of this block: I mean, not _changing the color_, but deciding that the header shouldn't have any color at all. I've pushed at https://github.com/WordPress/gutenberg/pull/29731 an alternative to this PR that works well with text & background that should help illustrate this point.

- **Nested content**. What should the colors be if the calendar is not a top-level block (it's part of a container block such as group or columns, or perhaps part of a template part, etc)? It seems like for this approach to work well we have to define first which "containers" create this kind of system. https://github.com/WordPress/gutenberg/issues/28163 and https://github.com/WordPress/gutenberg/issues/27233 could be useful for previous conversations.

Did you try to implement this with the style system (block supports / theme.json)? I'd be curious how far you can get. I can foresee some issues (the current block uses more colors than we have available via block supports).

**Comment by scruffian** [comment]
> What if the calendar block had different opinions about defaults

I don't see that this would be a problem. A block can choose to not specify any colors in which case it would simply inherit from its context. Alternatively it can implement its own colors as the calendar block currently does. This change opens up another option which I think would be very welcome to block developers - it gives blocks a way to know what the default colors of the site are and implement them in a flexible way.

> What should the colors be if the calendar is not a top-level block
This is a tricky problem, but it's one that exists already for blocks inside containers. I think the solution above could be useful in solving this though - we could redefine these root level variables in certain contexts. For example we could add:

```
.has-background-text-color {
--wp--styles--root--text: var(--wp--styles--root--background-color);
}
```

This would change the variable only in the scope defined by this selector.

> Did you try to implement this with the style system (block supports / theme.json)? 
I think the problem with this approach is that I still would have no way to access the default text and background colors for the block to set them as defaults. The only way to do it would be to expect them to be set via theme.json, which is possible for blocks supported in core, but it's not practical for themes to set defaults for all possible third party blocks.

The advantage of the proposal is particularly for third party block developers - giving them a way to make their blocks look like the theme.

**Comment by carolinan** [comment]
I think the calendar block should have its own color and style options, like the table block.

**Comment by youknowriad** [comment]
This is interesting proposal :) 

So like mentionned by @nosolosw is to that we initially started like that for Global styles:

 1- Generate a CSS variable for each style config
 2- Ask block authors to use these variables in their CSS appropriately.
 3- Do not generate any direct style.

We quickly realized that it doesn't work though because of the CSS variable inheritence: when you apply a background to a container, all blocks inside it that support this variable (think  buttons, tables...) get their background change in addition to their container, which is not the expected result. This is why we changed the approach, we decided to avoid step 1 and 2 above and just do the step 3 which is: generate the styles directly based on the config. This approach proved to work in a lot of situations just fine.

But what you're showing in this PR that there might be still value in the original approach if it's only "opt-in" per blocks and not the only way to consume global styles.

What you're doing in this PR  is:

For each global style config, generate two things: an actual style (like color: red) and  a CSS variable definition (--my-var: red). In some situtations, the CSS variable is used, in others we just rely on the actual style and in others, we need both (containers like the root block or any group block come to my mind when I think about blocks that need both definitions)

So this PR is impactful, it changes Global styles conceptually a bit by introducing "two modes" of application of global styles. I'm not against it personally but we'd have to think it properly:

 - Do we want to always apply the two modes to all declarations? (maybe that's fine and sufficient)
 - Do we want to make the CSS variables mode specific to some contexts but not others (For instance in this PR, I see it being useful for the root context, group block and the actual calendar block) and potentially not useful for others? Do we want an API for block to say whether they allow CSS variables in their global style context?
 - Are there any other heuristics like the property itself (for example colors could generate vars while font sizes or something else not, this is of course not a good example, just useful to express the idea)

I'd love more thoughts from @mtias @mcsf and if we ever we decide to go into this direction, we'd have to make sure to clarify the behavior properly in our docs...

**Comment by MaggieCabrera** [comment]
I find the most important part of this PR to be that we are giving both plugin and theme authors access to the variables that otherwise they didn't have access to (compared to how they can access others such as blocks') that are very important for the theme since they are on the root level. 
I understand the problems with inheritance and maybe not all blocks should be making use of these variables, but I see great value in at least having them available.

**Comment by nosolosw** [comment]
I left my brain to think about this while sleeping to see how we could address the issues above. Didn't have a lot of time to sum up my thoughts, so my braindump is a bit wordy. TLDR ― I think the core idea has the potential to enable something we aren't able to provide at the moment: maintaining contextual design links, which touches on things we were talking about at https://github.com/WordPress/gutenberg/issues/27233

Let's say we have a calendar as a top-level block (within `root`) and another calendar within a group. We also have this theme.json:

```json
"root": {
  "text": "var(--wp--preset--color--red)"
},
"core/group": {
  "text": "var(--wp--preset--color--blue)"
},
"core/calendar": {
  "background": "var(--wp--preset--color--red)"
}
```

At the moment:

1. If the user changes the *value* of the presets: make `--wp--preset--color--red` darker and `--wp--preset--color--blue` lighter. The top-level calendar block will react to the root change. We don't have a way to wire-up the calendar within the group to react to the environment it lives within. This can only happen at the global level (gs sidebar).
2. If the user changes the *value* of the calendar block. The intention is that the block has a color of its own, not related to the parents. Doesn't need any contextual information. This works fine today. This can happen at the global level (GS sidebar) or at the local level (block sidebar).
3. If the user changes the *value* of the container blocks (for example, sets the color of the group to `yellow` and the color of the root to `pink`). Both calendars will stay as they are, they won't change according to the environment they live in.
	1. This can happen either globally (GS sidebar).
	2. This can happen locally (block sidebar).

As Riad suggested, what if _all containers_ (or blocks that opt-in via an API) output both styles and CSS vars. theme.json input:

```
"root": {
  "text": "var(--wp--preset--color--red)"
},
"core/group": {
  "text": "var(--wp--preset--color--blue)"
},
"core/calendar": {
  "background": "var(--wp--style--color--text)" // NOTE THIS VARIABLE.
}
```

The CSS output will be:

```css
:root {
  color: var(--wp--preset--color--red);
  --wp--style--color--text: var(--wp--preset--color--red);
}
.wp-block-group {
  color: var(--wp--preset--color--blue);
  --wp--style--color--text: var(--wp--preset--color--blue);
}

.wp-block-calendar{
  color: var(--wp--style--color--text); // NOTE THIS VARIABLE.
}
```

With this CSS we can make both 1 and 3.1 work because the value of the variable used by the calendar depends on the environment (root, group, etc) but it's agnostic as to which environment it lives in.

We can do this with one variable per each style property we expose via theme.json. *If* we wanted all properties, this would be the whole list:

```css
--wp--style--border--color
--wp--style--border--radius
--wp--style--border--style
--wp--style--border--width
--wp--style--color--text
--wp--style--color--background
--wp--style--color--link
--wp--style--color--gradients
--wp--style--spacing--padding
--wp--style--typography--font-size
--wp--style--typography--font-family
--wp--style--typography--font-weight
--wp--style--typography--line-height
--wp--style--typography--text-decoration
--wp--style--typography--text-transform
```

However, we still need to make 3.2 work as well.

To make it backward-compatible, we need an approach that doesn't serialize anything to the DOM because we can't rewrite old posts. What if we use the preset classes to set the variables as well? Example:

What we do now:

```css
.has-red-color {
  color: red;
}

// At the moment, we only enqueue this if the block
// defines a preset different than the defaults.
.wp-block-columns.has-red-color {
  color: #c6352e;
}
```

What we could do:

```css
.has-red-color,
.wp-block-group.has-red-color,
.wp-block-columns.has-red-color {
  color: red;
  --wp--style--color--text: red;
}

// We still do this if the block
// defines its own presets.
// It'll overrride the above's ruleset
// if it's present.
.wp-block-columns.has-red-color {
  color: #c6352e;
  --wp--style--color--text: #c6352e;
}
```

This would work for 3.2, but not fully: what if the user selects for the container a custom color (which doesn't attach the preset classes to the block but uses inline styles instead). For new blocks, we could add the CSS variable as inline styles for the container, but this won't work to address old posts.

So, this is how far I've gotten to develop this idea. The issue I am not sure how to address is custom colors set at the block level for old posts. But the rest seems that it could work really nicely and reduces the need for "nested contexts" we were talking about in https://github.com/WordPress/gutenberg/issues/27233 Perhaps we only need to be able to add template parts to the equation. I see a couple of ways to do it, but don't want to divert the conversation from this.

**Comment by nosolosw** [comment]
edited my comment :point_up: to fix a couple of typos

**Comment by oandregal** [comment]
https://github.com/WordPress/gutenberg/issues/29891 may be relevant to this discussion. In particular, as per [design mockups](https://github.com/WordPress/gutenberg/issues/27473#issuecomment-796773866), there's going to be a background element blocks can tap into. The specific implementation of elements (chaining selectors, CSS Custom Properties, etc) is going to be defined as we implement each.

**Comment by youknowriad** [comment]
@scruffian @nosolosw I've been thinking about this PR again and I think we should do it, but I also think that we should be doing it with the following rules:

 - On root level generate both CSS variables and regular styles
 - On block level (or block context) which means whenever someone writes this in "theme.json"

```
"my/block": {
    "color": { "text": "red" }
}
```

In this case, we should check the config of the block itself:  

 - if we do not find the block or do not find any config, generate regular styles (no css variables)
 - if we find a `style` (new top level property that is different than "supports" because I think these two thing are separate) config in `block.json` like that:

```
"style": {
   "color": {
      "text": "variable" // values here can be: "variable", "default", or [ "variable", "default"]
   }
}
```

generate the style according to the config.
 
- block supports (hooks) can also use this style to generate the styles properly

**Reasoning**

 - This allows us to avoid the "skipSerialization" flag we're introducing to different flags and blocks and instead have everything rely on the CSS variable for these cases. (for instance setting the color of the button block)
 - block supports is a way to add features to blocks but doesn't map 1-1 to theme.json
 - This proposal reintroduces the behavior that was originally used for Global styles but that proved to be non useful and that breaks if applied consistently to all styles but it does solve some use-cases where regular styles applied to the wrapper don't work.

I'd love your thoughts on that  @mtias @mcsf @scruffian @nosolosw @ockham @gziolo

**Comment by gziolo** [comment]
> This allows us to avoid the "skipSerialization" flag we're introducing to different flags and blocks and instead have everything rely on the CSS variable for these cases. (for instance setting the color of the button block)

When talking privately with @nosolosw I asked if we could use CSS variables instead of "skipSerialization" flag and he mentioned some blockers. If what you are proposing resolves them, then I'm all for the proposed approach. Otherwise, we will be forced to find a way to pass all the generated styles and class names to the element that isn't a wrapper. That would mean also that developers would have to explicitly use some new API method that is responsible for that. If CSS variables is an option then we should make sure it is used as the API you outlined is far simpler than the amount of work to understand and properly apply "skipSerialization".

**Comment by scruffian** [comment]
I agree this is much simpler than skipSerialization. It seems to give a lot of power while remaining very simple.

**Comment by ockham** [comment]
> - if we find a `style` (new top level property that is different than "supports" because I think these two thing are separate) config in `block.json` like that:
>
> ```
> "style": {
>   "color": {
>      "text": "variable" // values here can be: "variable", "default", or [ "variable", "default"]
>   }
>}
>```
> generate the style according to the config.

Apologies if this is obvious (I'm still trying to wrap my head around a number of these concepts), but what happens if the value is `"default"` or `[ "variable", "default"]`?

**Comment by youknowriad** [comment]
> Apologies if this is obvious (I'm still trying to wrap my head around a number of these concepts), but what happens if the value is "default" or [ "variable", "default"]?

Just the regular global styles behavior, so in this particular case it results in `.my-block { color: 'red' }` if it's in theme.json and just an inline style `color: red` if it's a block attribute value.

**Comment by ockham** [comment]
I've been staring at this for a while, and I'm afraid I still don't really grasp the proposal 😅  

One thing I've noticed is that this allows for a rather large number of different combinations of `theme.json` and `block.json` settings, which can generate quite different outputs in a site-wide generated CSS file (which is created based on settings in both files?), and inline style (block `style` attributes/markup?).

Would it be possible to list all different combinations, and spell out what the "global" generated CSS, and block attributes/markup would look like? I'm afraid I can't really weigh in on this otherwise.

One off-the-cuff question would be: If this affects block attributes/markup, won't it lead to deprecations upon theme switch?

**Comment by youknowriad** [comment]
@ockham would this comment help with the understanding of the proposal here https://github.com/WordPress/gutenberg/pull/26752#discussion_r604681082

**Comment by youknowriad** [comment]
I've spent some time thinking about this PR and the proposals above more and I came up to the conclusion that if we do it, it will suffer from the same issues that forced us to drop the CSS variables behavior of global styles in the first place. Let me clarify:

Imagine there's a block (say table) that relies on the background color and apply it to its cells like so:

```
table td {
   background-color: var(--wp-style-color-background);
}
```

And in theme.json we define a background color in the root

```
color: { background: 'red' }
```

What happens here is that we'll have:

 - Background color red applied to the canvas (root)
 - CSS variables defined at the root level
 - Background color applied to all table cells ---> this is not expected IMO (imagine you applied the background color to an intermediary container block)

In other words, this proposal only doesn't work for properties that doesn't cascade and only works for the ones that cascade (text color) and even for these, why not just rely on the regular cascade.

So I'm leaning towards avoiding this proposal entirely at least for v1 of Global styles.

**Comment by scruffian** [comment]
@youknowriad I'm not exactly sure why you think this is a bad idea, but I've removed the change to the calendar block in case that was your concern. Please could you give more details about why you don't like this approach?

From a themes perspective it will be very useful to have access to colors like the background that the user has set, as we might want to use it in places that won't cascade naturally.

**Comment by youknowriad** [comment]
> the background that the user has set, as we might want to use it in places that won't cascade naturally.

I tried explaining here https://github.com/WordPress/gutenberg/pull/29714#issuecomment-813910350 in general trying to benefit from a cascade like behavior for things that don't cascade can cause unexpected results.

Would you mind giving an example?

**Comment by scruffian** [comment]
> Would you mind giving an example?

Sure. So lets say I want to style table blocks in my theme, and I want the header of the table to contrast with the body like this:

Screenshot 2021-04-30 at 15 33 50

If I use CSS to achieve this, then if the user changes their background color to black using Global Styles, the header of the table won't be high contrast like this anymore.

Screenshot 2021-04-30 at 15 36 36

However if we had these values as CSS variables then I could use those variables in my CSS. Then when the user changed their colors the table would look like this:

Screenshot 2021-04-30 at 15 39 41

Does that help?

**Comment by youknowriad** [comment]
@scruffian what happens if the table is inside a group block and that the user changes the background and text color for that particular group?

**Comment by scruffian** [comment]
The table would still inherit the values from root, so it might look like this:
Screenshot 2021-04-30 at 16 01 39

**Comment by youknowriad** [comment]
@scruffian I don't understand why it inherits just the top level background/text colors? Imagine it's the sidebar where we invert the colors or there's a frame around the content where we invert the colors?

**Comment by scruffian** [comment]
If it were to inherit the rules from the group block then it would look like this:
Screenshot 2021-04-30 at 16 23 03

**Comment by youknowriad** [comment]
Ok understood, so we're saying this:

 - Each container should define variables for their properties.
 - Blocks can use these variables to style things in their markup.

unlike my example here https://github.com/WordPress/gutenberg/pull/29714#issuecomment-813910350 you're not using the "background color" as a background but more as a "text color". It does seem very weird to me in the first place because why would something change for me as a user when setting a "background color" to a specific area. That said, maybe we can be liberal here and accept that themes  can do this (not blocks).

The issue becomes that we'll be generating a lot of variables assignments, we can try it though and see. cc @nosolosw @mtias @mcsf 

Also, how do we know which blocks generate these variables and where will we do it?

**Comment by scruffian** [comment]
I think that setting variables for root will be useful. I'm less sure that we need to do the same for every container. It's probably sufficient most of the time to simply have the root variables.

**Comment by scruffian** [comment]
Closing in favour of https://github.com/WordPress/gutenberg/pull/39432
